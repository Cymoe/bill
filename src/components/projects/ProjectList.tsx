import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { MoreVertical } from 'lucide-react';
import { db } from '../../lib/database';
import type { Tables } from '../../lib/database';
import { DashboardLayout } from '../layouts/DashboardLayout';
import { PageHeader } from '../common/PageHeader';
import { NewButton } from '../common/NewButton';
import { TableSkeleton } from '../skeletons/TableSkeleton';
import { Dropdown } from '../common/Dropdown';
import { formatCurrency } from '../../utils/format';

type Project = Tables['projects'];

export const ProjectList: React.FC = () => {
  const navigate = useNavigate();
  const [projects, setProjects] = useState<Project[]>([]);
  const [loading, setLoading] = useState(true);
  const [selectedRows, setSelectedRows] = useState<string[]>([]);

  useEffect(() => {
    const fetchProjects = async () => {
      try {
        const projectsData = await db.projects.list();
        setProjects(projectsData);
      } catch (error) {
        console.error('Error fetching projects:', error);
      } finally {
        setLoading(false);
      }
    };

    fetchProjects();
  }, []);

  const getStatusColor = (status: Project['status']) => {
    switch (status) {
      case 'active':
        return 'bg-[#388E3C] text-[#6BFF90]';
      case 'completed':
        return 'bg-[#0D47A1] text-[#336699]';
      case 'on-hold':
        return 'bg-[#FFA726] bg-opacity-20 text-[#F9D71C]';
      case 'cancelled':
        return 'bg-[#D32F2F] bg-opacity-20 text-[#D32F2F]';
      default:
        return 'bg-[#333333] text-[#9E9E9E]';
    }
  };

  // Bulk select logic
  const allSelected = projects.length > 0 && selectedRows.length === projects.length;
  const toggleSelectAll = () => {
    if (allSelected) setSelectedRows([]);
    else setSelectedRows(projects.map(project => project.id));
  };
  const toggleSelectRow = (id: string) => {
    setSelectedRows((prev) => prev.includes(id) ? prev.filter(x => x !== id) : [...prev, id]);
  };

  const rowDropdownItems = (project: Project) => [
    {
      label: 'View Details',
      onClick: () => navigate(`/projects/${project.id}`),
      className: 'font-bold text-white',
    },
    {
      label: '',
      onClick: () => {},
      className: 'pointer-events-none border-t border-[#35384A] my-1',
    },
    {
      label: 'Edit Project',
      onClick: () => navigate(`/projects/${project.id}/edit`),
      className: 'font-bold text-white',
    },
    {
      label: '',
      onClick: () => {},
      className: 'pointer-events-none border-t border-[#35384A] my-1',
    },
    {
      label: 'Delete',
      onClick: () => {/* delete logic placeholder */},
      className: 'font-bold text-[#D32F2F]',
    },
  ];

  // Project summary statistics
  const activeProjects = projects.filter(project => project.status === 'active');
  const completedProjects = projects.filter(project => project.status === 'completed');
  const onHoldProjects = projects.filter(project => project.status === 'on-hold');
  
  const totalBudget = projects.reduce((sum, project) => sum + project.budget, 0);
  const activeBudget = activeProjects.reduce((sum, project) => sum + project.budget, 0);

  return (
    <DashboardLayout>
      <PageHeader
        hideTitle={true}
      />
      
      {/* Project summary cards */}
      <div className="hidden md:flex gap-0 mb-6">
        {/* Total Projects */}
        <div className="flex-1 border border-[#35384A] border-r-0 p-4 flex flex-col justify-center min-w-[180px]">
          <span className="text-sm text-gray-400 mb-1 font-['Roboto']">Total Projects</span>
          <span className="text-2xl font-bold text-white font-['Roboto_Condensed']">{projects.length}</span>
          <span className="text-xs text-gray-500 font-['Roboto']">Total Budget: {formatCurrency(totalBudget)}</span>
        </div>
        {/* Active Projects */}
        <div className="flex-1 border border-[#35384A] border-r-0 p-4 flex flex-col justify-center min-w-[180px]">
          <span className="text-sm text-gray-400 mb-1 font-['Roboto']">Active Projects</span>
          <span className="text-2xl font-bold text-[#6BFF90] font-['Roboto_Condensed']">{activeProjects.length}</span>
          <span className="text-xs text-gray-500 font-['Roboto']">Budget: {formatCurrency(activeBudget)}</span>
        </div>
        {/* On Hold */}
        <div className="flex-1 border border-[#35384A] border-r-0 p-4 flex flex-col justify-center min-w-[180px]">
          <span className="text-sm text-gray-400 mb-1 font-['Roboto']">On Hold</span>
          <span className="text-2xl font-bold text-[#F9D71C] font-['Roboto_Condensed']">{onHoldProjects.length}</span>
          <span className="text-xs text-gray-500 font-['Roboto']">{formatCurrency(onHoldProjects.reduce((sum, p) => sum + p.budget, 0))}</span>
        </div>
        {/* Completed */}
        <div className="flex-1 border border-[#35384A] p-4 flex flex-col justify-center min-w-[180px]">
          <span className="text-sm text-gray-400 mb-1 font-['Roboto']">Completed</span>
          <span className="text-2xl font-bold text-[#336699] font-['Roboto_Condensed']">{completedProjects.length}</span>
          <span className="text-xs text-gray-500 font-['Roboto']">{formatCurrency(completedProjects.reduce((sum, p) => sum + p.budget, 0))}</span>
        </div>
      </div>

      <div className="flex justify-between items-center mb-4">
        <h2 className="text-xl font-bold text-white font-['Roboto_Condensed'] uppercase">Projects</h2>
        <NewButton 
          onClick={() => navigate('/projects/new')} 
          label="New Project"
          color="yellow"
          className="rounded-[4px] font-bold font-['Roboto']"
        />
      </div>

      {loading ? (
        <TableSkeleton rows={5} columns={6} />
      ) : (
        <div className="bg-[#121212] rounded-[4px] shadow overflow-hidden border border-[#333333]">
          <div className="max-h-[calc(100vh-240px)] overflow-y-auto">
            <table className="min-w-full bg-[#121212]">
              <thead>
                <tr className="bg-[#1E1E1E] sticky top-0 z-10">
                  <th className="w-12 px-3 py-4 align-middle">
                    <div className="flex items-center h-full">
                      <input
                        type="checkbox"
                        className="form-checkbox h-4 w-4 text-[#336699] bg-transparent border-[#555555] rounded-sm focus:ring-[#0D47A1] focus:ring-opacity-40"
                        checked={allSelected}
                        onChange={toggleSelectAll}
                      />
                    </div>
                  </th>
                  <th className="text-left px-3 py-4 font-bold text-white font-['Roboto_Condensed'] uppercase">NAME</th>
                  <th className="text-left px-3 py-4 font-bold text-white font-['Roboto_Condensed'] uppercase">STATUS</th>
                  <th className="text-left px-3 py-4 font-bold text-white font-['Roboto_Condensed'] uppercase">BUDGET</th>
                  <th className="text-left px-3 py-4 font-bold text-white font-['Roboto_Condensed'] uppercase">START DATE</th>
                  <th className="text-left px-3 py-4 font-bold text-white font-['Roboto_Condensed'] uppercase">END DATE</th>
                  <th className="w-8 px-3 py-4"></th>
                </tr>
              </thead>
              <tbody className="divide-y divide-[#333333]">
                {projects.map((project) => (
                  <tr
                    key={project.id}
                    className={`transition-colors ${selectedRows.includes(project.id) ? 'bg-[#1E1E1E]' : 'hover:bg-[#1E1E1E]'} cursor-pointer`}
                    onClick={() => toggleSelectRow(project.id)}
                  >
                    <td className="px-3 py-4 align-middle">
                      <div className="flex items-center h-full" onClick={(e) => e.stopPropagation()}>
                        <input
                          type="checkbox"
                          className="form-checkbox h-4 w-4 text-[#336699] bg-transparent border-[#555555] rounded-sm focus:ring-[#0D47A1] focus:ring-opacity-40"
                          checked={selectedRows.includes(project.id)}
                          onChange={() => toggleSelectRow(project.id)}
                        />
                      </div>
                    </td>
                    <td className="px-3 py-4 font-medium text-white font-['Roboto']">{project.name}</td>
                    <td className="px-3 py-4">
                      <div className="flex items-center gap-2">
                        <span className="inline-flex items-center">
                          <span className="w-5 h-5 bg-[#333333] rounded-[4px] flex items-center justify-center mr-2">
                            <span className={`block w-2 h-2 ${getStatusColor(project.status).split(' ')[1]} rounded-[4px]`}></span>
                          </span>
                          <span className="text-white text-sm font-['Roboto'] capitalize">{project.status}</span>
                        </span>
                      </div>
                    </td>
                    <td className="px-3 py-4 text-white font-['Roboto_Mono'] font-medium">{formatCurrency(project.budget)}</td>
                    <td className="px-3 py-4 text-white font-['Roboto']">{new Date(project.start_date).toLocaleDateString()}</td>
                    <td className="px-3 py-4 text-white font-['Roboto']">{new Date(project.end_date).toLocaleDateString()}</td>
                    <td className="px-3 py-4 text-right" onClick={(e) => e.stopPropagation()}>
                      <Dropdown
                        trigger={
                          <button className="text-[#9E9E9E] hover:text-[#F9D71C]">
                            <MoreVertical className="w-5 h-5" />
                          </button>
                        }
                        items={rowDropdownItems(project)}
                      />
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
        </div>
      )}
    </DashboardLayout>
  );
};

export default ProjectList;
