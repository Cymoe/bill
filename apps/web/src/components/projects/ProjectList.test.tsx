/// <reference types="@testing-library/jest-dom" />
import React, { ReactNode } from 'react';
import { render, screen } from '@testing-library/react';
import { ProjectList } from './ProjectList';

// Mock react-router-dom
jest.mock('react-router-dom', () => ({
  useNavigate: () => jest.fn(),
}));

// Mock db
jest.mock('../../lib/database', () => ({
  db: {
    projects: {
      list: jest.fn().mockResolvedValue([
        {
          id: '1',
          name: 'Test Project',
          status: 'active',
          budget: 10000,
          start_date: '2024-01-01T00:00:00Z',
          end_date: '2024-12-31T00:00:00Z',
        },
        {
          id: '2',
          name: 'Completed Project',
          status: 'completed',
          budget: 20000,
          start_date: '2023-01-01T00:00:00Z',
          end_date: '2023-12-31T00:00:00Z',
        },
      ]),
    },
  },
}));

// Silence DashboardLayout
jest.mock('../layouts/DashboardLayout', () => ({
  DashboardLayout: ({ children }: { children: ReactNode }) => <div data-testid="dashboard-layout">{children}</div>,
}));

// Silence PlusIcon
jest.mock('./ProjectList', () => {
  const actual = jest.requireActual('./ProjectList');
  return {
    ...actual,
    PlusIcon: () => <svg data-testid="plus-icon" />,
  };
});

describe('ProjectList (Design System)', () => {
  it('renders the projects table with design system components', async () => {
    render(<ProjectList />);
    // Wait for projects to load
    expect(await screen.findByText('Test Project')).toBeInTheDocument();
    expect(screen.getByText('Completed Project')).toBeInTheDocument();

    // Table headings
    expect(screen.getByText('Projects')).toBeInTheDocument();
    expect(screen.getByText('Status')).toBeInTheDocument();
    expect(screen.getByText('Budget')).toBeInTheDocument();
    expect(screen.getByText('Start Date')).toBeInTheDocument();
    expect(screen.getByText('End Date')).toBeInTheDocument();

    // Buttons
    expect(screen.getByText('New Project')).toBeInTheDocument();
    expect(screen.getAllByText('Edit').length).toBeGreaterThan(0);

    // Status pills
    expect(screen.getByText('active')).toBeInTheDocument();
    expect(screen.getByText('completed')).toBeInTheDocument();

    // Snapshot for visual regression
    expect(document.body).toMatchSnapshot();
  });
}); 