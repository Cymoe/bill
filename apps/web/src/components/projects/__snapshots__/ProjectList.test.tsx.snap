// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList (Design System) renders the projects table with design system components 1`] = `
<body>
  <div>
    <div
      data-testid="dashboard-layout"
    >
      <div
        class="max-w-7xl mx-auto"
      >
        <div
          class="flex justify-between items-center"
        >
          <div
            class="sm:flex-auto"
          >
            <h1
              class="text-xl font-medium text-white"
            >
              Projects
            </h1>
            <p
              class="mt-2 text-sm text-gray-400"
            >
              A list of all projects including their name, status, and key details.
            </p>
          </div>
          <div
            class="mt-4 sm:mt-0 sm:ml-16 sm:flex-none"
          >
            <button
              class="inline-flex items-center h-10 px-4 rounded-md text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
              type="button"
            >
              <svg
                class="h-5 w-5"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 4v16m8-8H4"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                class="ml-3"
              >
                New Project
              </span>
            </button>
          </div>
        </div>
        <div
          class="mt-8 -mx-4 overflow-x-auto sm:-mx-6 lg:-mx-8"
        >
          <div
            class="inline-block min-w-full py-2 align-middle md:px-6 lg:px-8"
          >
            <div
              class="overflow-hidden shadow rounded-lg bg-[#1e2532]"
            >
              <table
                class="min-w-full divide-y divide-gray-700"
              >
                <thead
                  class="bg-[#2a3441]"
                >
                  <tr>
                    <th
                      class="px-3 py-3.5 text-left text-sm font-medium text-gray-400"
                      scope="col"
                    >
                      Name
                    </th>
                    <th
                      class="px-3 py-3.5 text-left text-sm font-medium text-gray-400"
                      scope="col"
                    >
                      Status
                    </th>
                    <th
                      class="px-3 py-3.5 text-left text-sm font-medium text-gray-400"
                      scope="col"
                    >
                      Budget
                    </th>
                    <th
                      class="px-3 py-3.5 text-left text-sm font-medium text-gray-400"
                      scope="col"
                    >
                      Start Date
                    </th>
                    <th
                      class="px-3 py-3.5 text-left text-sm font-medium text-gray-400"
                      scope="col"
                    >
                      End Date
                    </th>
                    <th
                      class="relative py-3.5 pl-3 pr-4 sm:pr-6"
                      scope="col"
                    >
                      <span
                        class="sr-only"
                      >
                        Actions
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="divide-y divide-gray-700 bg-transparent"
                >
                  <tr
                    class="cursor-pointer hover:bg-[#2a3441] transition-colors duration-200"
                  >
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm text-white"
                    >
                      Test Project
                    </td>
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm"
                    >
                      <span
                        class="inline-flex rounded-full px-2 text-xs font-semibold leading-5 bg-[#6BFF90] bg-opacity-8 text-[#6BFF90]"
                      >
                        active
                      </span>
                    </td>
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm text-white"
                    >
                      $
                      10,000
                    </td>
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm text-white"
                    >
                      1/1/2024
                    </td>
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm text-white"
                    >
                      12/31/2024
                    </td>
                    <td
                      class="relative whitespace-nowrap py-4 pl-3 pr-4 text-right text-sm font-medium sm:pr-6"
                    >
                      <button
                        class="text-indigo-400 hover:text-indigo-300 transition-colors duration-200"
                      >
                        Edit
                      </button>
                    </td>
                  </tr>
                  <tr
                    class="cursor-pointer hover:bg-[#2a3441] transition-colors duration-200"
                  >
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm text-white"
                    >
                      Completed Project
                    </td>
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm"
                    >
                      <span
                        class="inline-flex rounded-full px-2 text-xs font-semibold leading-5 bg-[#FF3B30] bg-opacity-8 text-[#FF3B30]"
                      >
                        completed
                      </span>
                    </td>
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm text-white"
                    >
                      $
                      20,000
                    </td>
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm text-white"
                    >
                      1/1/2023
                    </td>
                    <td
                      class="whitespace-nowrap px-3 py-4 text-sm text-white"
                    >
                      12/31/2023
                    </td>
                    <td
                      class="relative whitespace-nowrap py-4 pl-3 pr-4 text-right text-sm font-medium sm:pr-6"
                    >
                      <button
                        class="text-indigo-400 hover:text-indigo-300 transition-colors duration-200"
                      >
                        Edit
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
